File: elasticache.ts
import * as autoscaling from 'aws-cdk-lib/aws-autoscaling';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as elasticache from 'aws-cdk-lib/aws-elasticache';
import * as elbv2 from 'aws-cdk-lib/aws-elasticloadbalancingv2';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as cdk from 'aws-cdk-lib/core';
import { Construct } from 'constructs';


export interface StickySessionsMultiRegionProps extends cdk.StackProps {
    readonly CertificateArn: string,
    readonly HealthCheckPath: string;
    readonly HealthCheckPort: string;
    readonly HealthCheckHttpCodes: string;
}

export class StickySessionsMultiRegionStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props: StickySessionsMultiRegionProps) {
        super(scope, id, props);

        // Create a VPC
        const vpc = new ec2.Vpc(this, 'MyVPC', {
            maxAzs: 2, // Maximum Availability Zones
            subnetConfiguration: [
                {
                    cidrMask: 24,
                    name: 'Public',
                    subnetType: ec2.SubnetType.PUBLIC,
                },
                {
                    cidrMask: 24,
                    name: 'Private',
                    subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,
                },
            ]
        });

        // Security group for ElastiCache
        const cacheSecurityGroup = new ec2.SecurityGroup(this, 'CacheSecurityGroup', {
            vpc: vpc,
            description: 'Security group for ElastiCache',
        });

        // Allow inbound traffic from EC2 instance
        cacheSecurityGroup.addIngressRule(ec2.Peer.ipv4(vpc.vpcCidrBlock), ec2.Port.tcp(6379), 'Allow inbound from VPC');

        // Create ElastiCache cluster
        const cacheCluster = new elasticache.CfnCacheCluster(this, 'MyCacheCluster', {
            cacheNodeType: 'cache.t2.micro', // Adjust as per your requirements
            engine: 'redis', // Adjust as per your requirements
            numCacheNodes: 1, // Adjust as per your requirements
            vpcSecurityGroupIds: [cacheSecurityGroup.securityGroupId], // You may need to provide security group IDs
            cacheSubnetGroupName: 'my-cache-subnet-group', // You may need to provide a subnet group name
            clusterName: 'my-cache-cluster', // Adjust as per your requirements
        });

        // Create Redis Subnet Group based on the VPC Private Subnets
        const redisSubnetGroup = new elasticache.CfnSubnetGroup(
            this,
            `redisSubnetGroup`,
            {
                description: "Subnet group for the redis cluster",
                subnetIds: vpc.privateSubnets.map((ps) => ps.subnetId),
                cacheSubnetGroupName: "GT-Redis-Subnet-Group",
            }

        );
        cacheCluster.addDependency(redisSubnetGroup)

        // Create an Application Load Balancer
        const alb = new elbv2.ApplicationLoadBalancer(this, 'MyALB', {
            vpc,
            internetFacing: true, // Enable internet-facing
            vpcSubnets: {
                subnetType: ec2.SubnetType.PUBLIC, // Use public subnets
            },
        });

        // Add a listener to the ALB
        const listener = alb.addListener('Listener', {
            port: 80,
            open: true,
        });

        const httpsListener = alb.addListener('ALBListenerHttps', {
            certificates: [elbv2.ListenerCertificate.fromArn(props.CertificateArn)],
            protocol: elbv2.ApplicationProtocol.HTTPS,
            port: 443,
            sslPolicy: elbv2.SslPolicy.TLS12
        })

        // Create an IAM role for EC2 instances
        const instanceRole = new iam.Role(this, 'InstanceRole', {
            assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
        });

        // Policy granting permission to access ElastiCache
        const policy = new iam.PolicyStatement({
            effect: iam.Effect.ALLOW,
            actions: [
                'elasticache:DescribeCacheClusters',
                'elasticache:DescribeCacheParameterGroups',
                'elasticache:DescribeCacheSecurityGroups',
                'elasticache:DescribeCacheSubnetGroups',
                'elasticache:DescribeEngineVersions',
                'elasticache:DescribeSnapshots',
                'elasticache:ListAllowedNodeTypeModifications',
                'elasticache:ListTagsForResource',
            ],
            resources: ['*'], // Adjust as per your requirements
        });

        // Attach policy to the role
        instanceRole.addToPolicy(policy);

        // Add permissions to the IAM role
        instanceRole.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonSSMManagedInstanceCore'));

        // Create an Autoscaling Group
        const asg = new autoscaling.AutoScalingGroup(this, 'MyAutoScalingGroup', {
            vpc,
            instanceType: ec2.InstanceType.of(ec2.InstanceClass.T2, ec2.InstanceSize.MICRO),
            machineImage: ec2.MachineImage.latestAmazonLinux2(),
            minCapacity: 2, // Minimum number of instances
            maxCapacity: 4, // Maximum number of instances
            desiredCapacity: 2, // Desired number of instances
            role: instanceRole, // Set the IAM role for instances
            vpcSubnets: {
                subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS, // Use private subnets
            },
        });


        // Add a target group for the ALB
        const targetGroup = new elbv2.ApplicationTargetGroup(this, 'MyTargetGroup', {
            port: 80,
            targets: [asg],
            vpc,
            targetType: elbv2.TargetType.INSTANCE,
            targetGroupName: 'MyTargetGroup', // Set the name of the target group
        });

        // Attach the target group to the listener
        listener.addTargetGroups('TargetGroup', {
            targetGroups: [targetGroup],
        });
        httpsListener.addTargets('TargetGroup', {
            port: 443,
            protocol: elbv2.ApplicationProtocol.HTTPS,
            targets: [asg],
            loadBalancingAlgorithmType: elbv2.TargetGroupLoadBalancingAlgorithmType.LEAST_OUTSTANDING_REQUESTS,
            healthCheck: {
                path: props.HealthCheckPath,
                port: props.HealthCheckPort,
                healthyHttpCodes: props.HealthCheckHttpCodes
            }
        })

        // Add the ALB security group to the Autoscaling Group
        alb.connections.allowFromAnyIpv4(ec2.Port.tcp(80), 'Allow inbound HTTP traffic from anywhere');
        alb.connections.allowFromAnyIpv4(ec2.Port.tcp(443), 'Allow inbound HTTPS traffic from anywhere');
        asg.connections.allowTo(alb, ec2.Port.tcp(80), 'Allow outbound HTTP traffic to ALB');
    }
}

File: golang/templates/elasticache.go
package templates

import (
	"github.com/aws/aws-cdk-go/awscdk/v2"
	"github.com/aws/aws-cdk-go/awscdk/v2/awsautoscaling"
	"github.com/aws/aws-cdk-go/awscdk/v2/awsec2"
	"github.com/aws/aws-cdk-go/awscdk/v2/awselasticache"
	"github.com/aws/aws-cdk-go/awscdk/v2/awselasticloadbalancingv2"
	"github.com/aws/aws-cdk-go/awscdk/v2/awsiam"
	"github.com/aws/jsii-runtime-go"
)

func CreateElastiCache() {
	// Create the app properly
	app := awscdk.NewApp(nil)

	// pass it as parameter or use this function as main
	stack := awscdk.NewStack(app, jsii.String("StickySessionStack"), &awscdk.StackProps{
		Env: &awscdk.Environment{
			Account: jsii.String("YOUR_ACCOUNT_ID"),
			Region:  jsii.String("YOUR_REGION"),
		},
	})

	// Create a VPC
	vpc := awsec2.NewVpc(stack, jsii.String("MyVPC"), &awsec2.VpcProps{
		MaxAzs: jsii.Number(2),
		SubnetConfiguration: &[]*awsec2.SubnetConfiguration{
			{
				CidrMask:   jsii.Number(24),
				Name:       jsii.String("Public"),
				SubnetType: awsec2.SubnetType_PUBLIC,
			},
			{
				CidrMask:   jsii.Number(24),
				Name:       jsii.String("Private"),
				SubnetType: awsec2.SubnetType_PRIVATE_WITH_EGRESS,
			},
		},
	})

	// Create a security group for Elasticache
	cacheSecurityGroup := awsec2.NewSecurityGroup(stack, jsii.String("CacheSecurityGroup"), &awsec2.SecurityGroupProps{
		Vpc:         vpc,
		Description: jsii.String("Security group for Elasticache"),
	})
	cacheSecurityGroup.AddIngressRule(awsec2.Peer_Ipv4(vpc.VpcCidrBlock()), awsec2.Port_Tcp(jsii.Number(6379)), jsii.String("Allow inbound from VPC"), nil)

	// Create the Elasticache cluster
	cacheCluster := awselasticache.NewCfnCacheCluster(stack, jsii.String("MyCacheCluster"), &awselasticache.CfnCacheClusterProps{
		CacheNodeType:       jsii.String("cache.t2.micro"),
		Engine:              jsii.String("redis"),
		NumCacheNodes:       jsii.Number(1),
		VpcSecurityGroupIds: jsii.Strings(*cacheSecurityGroup.SecurityGroupId()),
		ClusterName:         jsii.String("my-cache-cluster"),
	})

	subnets := vpc.PrivateSubnets()
	var subnetIds []*string
	for _, v := range *subnets {
		subnetIds = append(subnetIds, v.SubnetId())
	}
	// Create a subnet group for the Elasticache cluster
	redisSubnetGroup := awselasticache.NewCfnSubnetGroup(stack, jsii.String("RedisSubnetGroup"), &awselasticache.CfnSubnetGroupProps{
		Description:          jsii.String("Subnet group for the Redis cluster"),
		SubnetIds:            &subnetIds,
		CacheSubnetGroupName: jsii.String("Redis-Subnet-Group"),
	})
	cacheCluster.AddDependsOn(redisSubnetGroup)

	// Create an ALB in a public subnet
	alb := awselasticloadbalancingv2.NewApplicationLoadBalancer(stack, jsii.String("MyALB"), &awselasticloadbalancingv2.ApplicationLoadBalancerProps{
		Vpc:            vpc,
		InternetFacing: jsii.Bool(true),
		VpcSubnets: &awsec2.SubnetSelection{
			SubnetType: awsec2.SubnetType_PUBLIC,
		},
	})

	// Create an IAM role for EC2 instances
	instanceRole := awsiam.NewRole(stack, jsii.String("InstanceRole"), &awsiam.RoleProps{
		AssumedBy: awsiam.NewServicePrincipal(jsii.String("ec2.amazonaws.com"), nil),
	})
	instanceRole.AddManagedPolicy(awsiam.ManagedPolicy_FromAwsManagedPolicyName(jsii.String("AmazonSSMManagedInstanceCore")))

	// Create an Auto Scaling Group
	asg := awsautoscaling.NewAutoScalingGroup(stack, jsii.String("MyAutoScalingGroup"), &awsautoscaling.AutoScalingGroupProps{
		Vpc:             vpc,
		InstanceType:    awsec2.InstanceType_Of(awsec2.InstanceClass_T2, awsec2.InstanceSize_MICRO),
		MachineImage:    awsec2.MachineImage_LatestAmazonLinux2(nil),
		MinCapacity:     jsii.Number(2),
		MaxCapacity:     jsii.Number(4),
		DesiredCapacity: jsii.Number(2),
		Role:            instanceRole,
		VpcSubnets: &awsec2.SubnetSelection{
			SubnetType: awsec2.SubnetType_PRIVATE_WITH_EGRESS,
		},
	})

	targets := []awselasticloadbalancingv2.IApplicationLoadBalancerTarget{}
	targets = append(targets, asg)
	// Create a target group for the ALB
	awselasticloadbalancingv2.NewApplicationTargetGroup(stack, jsii.String("MyTargetGroup"), &awselasticloadbalancingv2.ApplicationTargetGroupProps{
		Port:       jsii.Number(80),
		Targets:    &targets,
		Vpc:        vpc,
		TargetType: awselasticloadbalancingv2.TargetType_INSTANCE,
	})

	// Attach the target group to the ALB listener
	listener := alb.AddListener(jsii.String("Listener"), &awselasticloadbalancingv2.BaseApplicationListenerProps{Port: jsii.Number(80)})
	httpslistener := alb.AddListener(jsii.String("HTTPSListener"), &awselasticloadbalancingv2.BaseApplicationListenerProps{Port: jsii.Number(443)})
	listener.AddTargets(jsii.String("TargetGroup"), &awselasticloadbalancingv2.AddApplicationTargetsProps{Targets: &targets, LoadBalancingAlgorithmType: awselasticloadbalancingv2.TargetGroupLoadBalancingAlgorithmType_LEAST_OUTSTANDING_REQUESTS})
	httpslistener.AddTargets(jsii.String("HTTPSTargetGroup"), &awselasticloadbalancingv2.AddApplicationTargetsProps{Targets: &targets})

	// Synth with options if necessary
	app.Synth(nil)
}
